contracts/roles.js

```diff
@@ -0,0 +1,917 @@
+    if (deltaToken) {
+      const role = await api.db.findOne('roles', { _id: candidate.roleId });
+      const inst = await api.db.findOne('instances', { _id: role.instanceId });
+      if (inst.voteToken !== deltaToken.symbol) return true;
+    }
+    candidate.approvalWeight = {
+      $numberDecimal: api.BigNumber(candidate.approvalWeight.$numberDecimal)
+        .plus(deltaApprovalWeight),
+    };
+    await api.db.update('candidates', candidate);
+
+    const role = await api.db.findOne('roles', { _id: candidate.roleId });
+    role.totalApprovalWeight = {
+      $numberDecimal: api.BigNumber(role.totalApprovalWeight.$numberDecimal)
+        .plus(deltaApprovalWeight),
+    };
+    await api.db.update('roles', role);
+
+    return true;
+  }
+  return false;
+}
+
+actions.createInstance = async (payload) => {
+  const {
+    voteToken, candidateFee, isSignedWithActiveKey,
+  } = payload;
+
+  const params = await api.db.findOne('params', {});
+  const { instanceCreationFee } = params;
+
+  // eslint-disable-next-line no-template-curly-in-string
+  const utilityTokenBalance = await api.db.findOneInTable('tokens', 'balances', { account: api.sender, symbol: "'${CONSTANTS.UTILITY_TOKEN_SYMBOL}$'" });
+
+  const authorizedCreation = api.BigNumber(instanceCreationFee).lte(0) || api.sender === api.owner
+    ? true
+    : utilityTokenBalance && api.BigNumber(utilityTokenBalance.balance).gte(instanceCreationFee);
+
+  if (api.assert(authorizedCreation, 'you must have enough tokens to cover the creation fee')
+    && api.assert(isSignedWithActiveKey === true, 'you must use a transaction signed with your active key')) {
+    if (candidateFee) {
+      if (!api.assert(typeof candidateFee === 'object'
+        && typeof candidateFee.method === 'string'
+        && FeeMethod.indexOf(candidateFee.method) !== -1
+        && typeof candidateFee.symbol === 'string'
+        && typeof candidateFee.amount === 'string' && api.BigNumber(candidateFee.amount).gte(0), 'invalid candidateFee properties')) return;
+      const feeTokenObj = await api.db.findOneInTable('tokens', 'tokens', { symbol: candidateFee.symbol });
+      if (!api.assert(feeTokenObj && api.BigNumber(candidateFee.amount).dp() <= feeTokenObj.precision, 'invalid candidateFee token or precision')) return;
+    }
+
+    const voteTokenObj = await api.db.findOneInTable('tokens', 'tokens', { symbol: voteToken });
+    if (!api.assert(voteTokenObj && voteTokenObj.stakingEnabled, 'voteToken must have staking enabled')) return;
+
+    const now = new Date(`${api.hiveBlockTimestamp}.000Z`);
+    const newInstance = {
+      voteToken,
+      candidateFee,
+      active: false,
+      creator: api.sender,
+      lastTickTime: now.getTime(),
+    };
+    const insertedInst = await api.db.insert('instances', newInstance);
+
+    if (api.sender !== api.owner
+        && api.sender !== 'null'
+        && api.BigNumber(instanceCreationFee).gt(0)) {
+      await api.executeSmartContract('tokens', 'transfer', {
```

Comment:
Although the possibility of this transfer failing is remote, it would be safe to transfer the fees before creating the new instance, and verify that the transfer succeeded.  See, for example, the technique used in the issue action of the nft contract.